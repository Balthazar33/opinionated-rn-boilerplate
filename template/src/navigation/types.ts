import {NavigationProp, ParamListBase, RouteProp, StackActionHelpers, StackNavigationState} from '@react-navigation/native';
import {NativeStackNavigationEventMap, NativeStackNavigationOptions} from 'react-native-screens/lib/typescript/native-stack/types';

export enum StackScreens {
  DummyScreen = 'DummyScreen',
  DummyScreenWithScroll = 'DummyScreenWithScroll',
  ApiCallScreen = 'ApiCallScreen',
  PokemonDetailScreen = 'PokemonDetailScreen',
}

export type NativeStackNavigationProp<
  ParamList extends ParamListBase,
  RouteName extends keyof ParamList = string,
  NavigatorID extends string | undefined = undefined
> = NavigationProp<
  ParamList,
  RouteName,
  NavigatorID,
  StackNavigationState<ParamList>,
  NativeStackNavigationOptions,
  NativeStackNavigationEventMap
> &
  StackActionHelpers<ParamList>;

export type NativeStackScreenProps<
  ParamList extends ParamListBase,
  RouteName extends keyof ParamList = string,
  NavigatorID extends string | undefined = undefined
> = {
  navigation: NativeStackNavigationProp<ParamList, RouteName, NavigatorID>;
  route: RouteProp<ParamList, RouteName>;
};

export type RootStackParamList = {
  /**
   * Add screens in the following format:
   * <screen name>: <the param(s) it expects>.
   * Specify 'undefined' if the screen does not expect any params.
   */
  DummyScreen: undefined;
  DummyScreenWithScroll: undefined;
  ApiCallScreen: undefined;
  PokemonDetailScreen: {name: string};
  // add more screens here...
};
