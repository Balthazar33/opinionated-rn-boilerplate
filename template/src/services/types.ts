import {AppDispatch} from '@redux/store.utils';

export interface ApiCallWithErrorHandling {
  apiCall: (params: any) => any;
  params: any;
  dispatch: AppDispatch;
}

export interface ApiCallResponse {
  data: null | any;
  error: {
    message?: string;
  };
}

interface Meta {
  headers?: any;
  status: number;
  statusText: string;
}

export interface CommonResponse {
  data: any;
  meta?: {
    response?: Meta | undefined;
  }; // API call meta data
}

export enum FetchErrorStatus {
  PARSING_ERROR = 'PARSING_ERROR',
}

export enum FetchErrorCodes {
  NOT_FOUND = 404,
  UNAUTHORIZED = 401,
  INCORRECT_INPUT = 400,
  SERVER_ERROR = 500,
}

export interface ApiError {
  data?: any;
  originalStatus: FetchErrorCodes | number;
  status: FetchErrorStatus | string;
}

export interface ServerError extends ApiCallWithErrorHandling {
  error: any;
  errorCode?: ErrorType;
}

export enum StatusCodes {
  ERROR400 = 400,
  ERROR401 = 401,
  ERROR404 = 404,
  ERROR500 = 500,
  ERROR300 = 300,
  SUCCESS200 = 200,
  SUCCESS201 = 201,
}

export type ErrorType =
  | StatusCodes.ERROR400
  | StatusCodes.ERROR401
  | StatusCodes.ERROR404
  | StatusCodes.ERROR500;
