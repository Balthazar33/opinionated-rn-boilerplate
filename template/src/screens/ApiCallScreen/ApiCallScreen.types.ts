import {RootStackParamList, StackScreens} from '@/navigation/types';
import {StackScreenProps} from '@react-navigation/stack';

export interface PokemonItemType {
  name: string;
  url: string;
}

export type ApiCallScreenProps = StackScreenProps<
  RootStackParamList,
  StackScreens.ApiCallScreen
>;
