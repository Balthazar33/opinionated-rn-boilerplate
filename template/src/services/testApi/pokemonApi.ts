import {createApi, fetchBaseQuery} from '@reduxjs/toolkit/query/react';
import {Endpoints} from './pokemonApi.endpoints';
import {FetchMethods} from '../constants';
import {CommonResponse} from '../types';
export const BASE_URL = 'https://pokeapi.co/api/v2/';

export interface GetAllPokemonQueryParams {
  limit?: number;
}

export const pokemonApi = createApi({
  reducerPath: 'pokeApi',
  baseQuery: fetchBaseQuery({baseUrl: BASE_URL}),
  endpoints: builder => ({
    // Get all pokemons
    getAllPokemon: builder.query<CommonResponse, GetAllPokemonQueryParams>({
      query: (params) => ({
        url: Endpoints.ALL,
        method: FetchMethods.GET,
        params: {
          ...params,
          limit: params?.limit ?? 10,
        },
      }),
      transformResponse: (response, meta, _) => ({data: response, meta}),
    }),
    // Get pokemon details by name
    getDetailsByName: builder.query({
      query: ({name}) => ({
        url: `${Endpoints.ALL}${name}`,
        method: FetchMethods.GET,
      }),
      transformResponse: (response, meta, _) => ({data: response, meta}),
    }),
    // The endpoint for the following POST request does not exist and is added merely as an example
    addPokemonByName: builder.mutation({
      query: body => ({
        url: Endpoints.ADDNEW,
        method: FetchMethods.POST,
        body,
      }),
      transformResponse: (response, meta, _) => ({data: response, meta}),
    }),
  }),
});

export const {
  useLazyGetAllPokemonQuery,
  useLazyGetDetailsByNameQuery,
  useAddPokemonByNameMutation,
} = pokemonApi;
