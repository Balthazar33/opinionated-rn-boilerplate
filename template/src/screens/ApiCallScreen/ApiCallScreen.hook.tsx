import React, {useCallback} from 'react';

import {PokemonItem} from './PokemonItem';
import {getAllPokemon} from './ApiCallScreen.utils';
import {useAppDispatch, useAppSelector} from '@/appRedux/store.utils';
import {PokemonItemType} from './ApiCallScreen.types';
import {GetAllPokemonQueryParams, useLazyGetAllPokemonQuery} from '@/services/testApi/pokemonApi';

export const useApiCallScreen = () => {
  const dispatch = useAppDispatch();
  const [trigger] = useLazyGetAllPokemonQuery();
  const {all}: {all: PokemonItemType[]} = useAppSelector(state => state.pokemon) || {};

  const renderItem = useCallback(
    ({item}: {item: PokemonItemType}) => (
      <PokemonItem {...item} />
    ),
    [],
  );

  const handleBtnPress = async () => {
    getAllPokemon<GetAllPokemonQueryParams>({dispatch, apiCall: trigger, params: {limit: 5}});
  };


  return {dispatch, all, trigger, renderItem, handleBtnPress};
};


