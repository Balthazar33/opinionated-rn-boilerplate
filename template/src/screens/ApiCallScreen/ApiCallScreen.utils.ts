import {setAll} from '@redux/slices/pokemonSlice';
import {AppDispatch} from '@redux/store.utils';
import {callApi} from '@services/apiCaller';
import {PokemonItemType} from './ApiCallScreen.types';

interface ApiCallParameters {
  dispatch: AppDispatch;
  apiCall: (params: any) => any;
  params: any;
}

interface PokemonResponse {
  results: PokemonItemType[];
}

export const getAllPokemon = async ({
  dispatch,
  apiCall,
  params,
}: ApiCallParameters) => {
  try {
    const response = await callApi({
      params,
      apiCall,
      dispatch,
    });
    // modify as per requirement
    if (response.error) {
      // throw error
      throw new Error(response.error.message);
    } else {
      // process data
      const responseData = response.data as PokemonResponse;
      dispatch(setAll(responseData?.results || []));
    }
  } catch (error) {
    // handle error
    console.log('Error', error);
  }
};
