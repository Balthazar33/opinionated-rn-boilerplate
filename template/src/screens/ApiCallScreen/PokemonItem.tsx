import React from 'react';
import {StyleSheet, Text} from 'react-native';

import {TouchableRipple} from 'react-native-paper';
import {sizer} from '../../utils/metrics';
import {ApiCallScreenProps} from './ApiCallScreen.types';
import {StackScreens} from '../../navigation/types';

interface PokemonItemProps {
  name: string;
  navigation: ApiCallScreenProps['navigation'];
}

export const PokemonItem = ({name, navigation}: PokemonItemProps) => {
  const handleNamePress = () =>
    navigation.navigate(StackScreens.PokemonDetailScreen, {name});

  return (
    <TouchableRipple style={style.wrapper} onPress={handleNamePress}>
      <Text>{name}</Text>
    </TouchableRipple>
  );
};

const style = StyleSheet.create({
  wrapper: {
    padding: sizer(12),
  },
});
