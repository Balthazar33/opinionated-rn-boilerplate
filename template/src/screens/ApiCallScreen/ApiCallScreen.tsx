import React, {useCallback} from 'react';
import {FlatList, StyleSheet} from 'react-native';

import {TouchableRipple} from 'react-native-paper';

import {sizer} from '@utils/metrics';
import {Strings} from '@utils/strings';
import {TestIds} from '@utils/test-ids';
import {PokemonItem} from './PokemonItem';
import {SCREEN_PADDING} from '@utils/constants';
import {getAllPokemon} from './ApiCallScreen.utils';
import BaseScreen from '../../containers/BaseScreen';
import {TextBold12, TextRegular12} from '@components/Typography';
import {useAppDispatch, useAppSelector} from '@redux/store.utils';
import {ApiCallScreenProps, PokemonItemType} from './ApiCallScreen.types';
import {useGetAllPokemonMutation} from '@/services/graphqlApi/pokeGraphApi';

export const ApiCallScreen = ({navigation}: ApiCallScreenProps) => {
  const dispatch = useAppDispatch();
  const {all}: {all: PokemonItemType[]} =
    useAppSelector(state => state.pokemon) || {};
  // extract api call from mutation
  const [pokemonApi] = useGetAllPokemonMutation();

  // getter & setter
  const handleBtnPress = () => {
    getAllPokemon({apiCall: pokemonApi, params: {count: 10}, dispatch});
  };

  const renderItem = useCallback(
    ({item}: {item: PokemonItemType}) => (
      <PokemonItem navigation={navigation} {...item} />
    ),
    [navigation],
  );

  return (
    <BaseScreen testID={TestIds.APICALL_SCREEN} style={style.screenStyle}>
      <TextRegular12>{Strings.apiCallScreen}</TextRegular12>
      <TouchableRipple
        onPress={handleBtnPress}
        style={style.buttonStyle}
        testID={TestIds.GET_ALL_POKEMON_BTN}>
        <TextBold12>{Strings.getAllPokemon}</TextBold12>
      </TouchableRipple>
      {/* List of pokemon */}
      <FlatList data={all} renderItem={renderItem} />
    </BaseScreen>
  );
};

const style = StyleSheet.create({
  screenStyle: {
    gap: sizer(16),
    padding: SCREEN_PADDING,
  },
  buttonStyle: {
    padding: sizer(16),
  },
});
