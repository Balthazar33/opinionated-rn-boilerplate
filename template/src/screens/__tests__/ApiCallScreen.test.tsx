jest.useFakeTimers();
import React from 'react';

import {setupServer} from 'msw/node';
import {http, HttpResponse, delay} from 'msw';
import {it, describe, expect} from '@jest/globals';
import '@testing-library/react-native/extend-expect';

import {ApiCallScreen} from '..';
import * as PokeApi from '../../services/testApi/pokemonApi';
import {renderWithProviders} from '../../utils/test-utils';
import {ApiCallScreenProps} from '../ApiCallScreen/ApiCallScreen.types';
import {BASE_URL} from '../../services/testApi/pokemonApi';
import {Endpoints} from '../../services/testApi/pokemonApi.endpoints';
import {TestIds} from '../../utils/test-ids';
import {fireEvent, waitFor, screen} from '@testing-library/react-native';

//Test Setup-----------------------------------------------------
const navigation: Partial<ApiCallScreenProps['navigation']> = {
  navigate: jest.fn,
};
const route: Partial<ApiCallScreenProps['route']> = {
  params: undefined,
};
export const handlers = [
  // Mocked API response
  http.get(`${BASE_URL}${Endpoints.ALL}`, async () => {
    await delay(150);
    return HttpResponse.json({
      results: [
        {
          name: 'bulbasaur',
          url: '',
        },
        {
          name: 'ivysaur',
          url: '',
        },
      ],
    });
  }),
];
const server = setupServer(...handlers);
//----------------------------------------------------------------
describe('ApiCallScreen', () => {
  // Enable API mocking before tests.
  beforeAll(() => server.listen());
  // Reset any runtime request handlers we may add during the tests.
  afterEach(() => server.resetHandlers());
  // Disable API mocking after the tests are done.
  afterAll(done => {
    server.close();
    done();
  });

  it('renders correctly', () => {
    const elementTree = renderWithProviders(
      <ApiCallScreen
        route={route as ApiCallScreenProps['route']}
        navigation={navigation as ApiCallScreenProps['navigation']}
      />,
    ).toJSON();
    expect(elementTree).toMatchSnapshot();
  });

  it('should fetch the list of pokemons on pressing the "Get all pokemon" button', async () => {
    const getAllSpyFn = jest.spyOn(PokeApi, 'useGetAllPokemonMutation');
    renderWithProviders(
      <ApiCallScreen
        route={route as ApiCallScreenProps['route']}
        navigation={navigation as ApiCallScreenProps['navigation']}
      />,
    );
    const getAllPokemonBtn = screen.getByTestId(TestIds.GET_ALL_POKEMON_BTN);
    await waitFor(() => {
      fireEvent(getAllPokemonBtn, 'press');
    });
    expect(getAllSpyFn).toHaveBeenCalled();
    expect(await screen.findByText('bulbasaur')).toBeOnTheScreen();
    expect(await screen.findByText('ivysaur')).toBeOnTheScreen();
  });
});
