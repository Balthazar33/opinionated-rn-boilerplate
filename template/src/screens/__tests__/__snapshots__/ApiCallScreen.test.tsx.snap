// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiCallScreen renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "transparent",
      "flex": 1,
    }
  }
  testID="AppSafeArea"
>
  <View
    style={
      [
        {
          "flex": 1,
        },
        {
          "gap": 32,
          "padding": 32,
        },
      ]
    }
    testID="ApiCallScreen"
  >
    <Text>
      Api Call Screen
    </Text>
    <View
      accessibilityState={
        {
          "disabled": false,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          false,
          undefined,
        ]
      }
    >
      <Text>
        Get All Pokemon
      </Text>
    </View>
    <RCTScrollView
      data={[]}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View />
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
